services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DefaultS3TransferManager.java

```diff
@@ -107,43 +113,111 @@ public Upload upload(UploadRequest uploadRequest) {
 
             CompletableFuture<PutObjectResponse> putObjFuture =
                 s3CrtAsyncClient.putObject(uploadRequest.putObjectRequest(), requestBody);
-            
+
             // Forward upload cancellation to CRT future
             CompletableFutureUtils.forwardExceptionTo(uploadFuture, putObjFuture);
 
-            CompletableFutureUtils.forwardTransformedResultTo(putObjFuture, uploadFuture, r -> CompletedUpload.builder()
-                                                                                                              .response(r)
-                                                                                                              .build());
+            CompletableFutureUtils.forwardTransformedResultTo(putObjFuture, uploadFuture,
+                                                              r -> CompletedUpload.builder()
+                                                                                  .response(r)
+                                                                                  .build());
         } catch (Throwable throwable) {
             uploadFuture.completeExceptionally(throwable);
         }
 
         return new DefaultUpload(uploadFuture, progressUpdater.progress());
     }
+    
+    @Override
+    public FileUpload uploadFile(UploadFileRequest uploadFileRequest) {
+        Validate.paramNotNull(uploadFileRequest, "uploadFileRequest");
+
+        AsyncRequestBody requestBody = AsyncRequestBody.fromFile(uploadFileRequest.source());
+
+        CompletableFuture<CompletedFileUpload> uploadFuture = new CompletableFuture<>();
+
+        TransferProgressUpdater progressUpdater = new TransferProgressUpdater(uploadFileRequest, requestBody);
+        progressUpdater.transferInitiated();
+        requestBody = progressUpdater.wrapRequestBody(requestBody);
+        progressUpdater.registerCompletion(uploadFuture);
+
+        try {
+            assertNotUnsupportedArn(uploadFileRequest.putObjectRequest().bucket(), "upload");
+
+            CompletableFuture<PutObjectResponse> putObjFuture =
+                s3CrtAsyncClient.putObject(uploadFileRequest.putObjectRequest(), requestBody);
+            
+            // Forward upload cancellation to CRT future
+            CompletableFutureUtils.forwardExceptionTo(uploadFuture, putObjFuture);
+
+            CompletableFutureUtils.forwardTransformedResultTo(putObjFuture, uploadFuture,
+                                                              r -> CompletedFileUpload.builder()
+                                                                                      .response(r)
+                                                                                      .build());
+        } catch (Throwable throwable) {
+            uploadFuture.completeExceptionally(throwable);
+        }
```

Comment:
Any chance that this lets the putObject keep running in the background? If so, can we cancel that as well?

Reply:
Hmm, I was thinking that line 151 (`forwardExceptionTo(..)`) would do this for us, but I'm not sure if `completeExceptionally(..)` will also implicitly `cancel(..)`?

If not, then perhaps we can extend `forwardExceptionTo(..)` to check for `CancellationException`, and attempt to `cancel(..)` first?

Reply:
If forwardExceptionTo throws an exception, it could, but I agree that's unlikely... unless someone comes along and adds more lines in there.

Reply:
This is the cancellation logic in S3 CRT client https://github.com/awslabs/aws-crt-java/blob/main/src/main/java/s3NativeClient/com/amazonaws/s3/S3NativeClient.java#L158 Looks like it checks if the future is cancelled or not.

Reply:
So I think it's probably a potential miss that `forwardExceptionTo(..)` is not forwarding/propagating cancellation. However, I'm not sure if it's safe to add that logic since the method is used in other places, including in generated clients. Alternatively, we create a new `forwardExceptionAndCancellationTo(..)` method, but it seems unfortunate to have to duplicate several of the existing methods, rather than just fix them. To prevent this PR from growing even further, I'd like to propose deferring this change to a separate, follow-up PR (which I will track).